esp-hal/src/clock/mod.rs

```diff
@@ -591,65 +599,135 @@ impl Clocks {
     }
 }
 
-/// Control the radio peripheral clocks
 #[cfg(any(bt, ieee802154, wifi))]
-#[instability::unstable]
-pub struct RadioClockController<'d> {
-    _rcc: crate::peripherals::RADIO_CLK<'d>,
+/// Tracks which modems currently request an active PHY clock.
+static PHY_CLOCK_REF_COUNTERS: critical_section::Mutex<Cell<u8>> =
+    critical_section::Mutex::new(Cell::new(0));
+
+// These functions are moved out of the trait to prevent monomorphization for
+// every modem clock controller. If that doens't really make sense, this can be
+// moved back.
+
+#[cfg(any(bt, ieee802154, wifi))]
+fn increase_phy_clock_ref_count_internal() {
+    critical_section::with(|cs| {
+        let phy_clock_ref_counter = PHY_CLOCK_REF_COUNTERS.borrow(cs);
+        let phy_clock_ref_count = phy_clock_ref_counter.get();
+
+        // If other modems have reference to the PHY clock, but this one doesn't, the
+        // PHY clock will just be set to enabled again, which shouldn't have any
+        // effect.
+        if phy_clock_ref_count == 0 {
+            clocks_ll::enable_phy(true);
+        }
+
+        phy_clock_ref_counter.set(phy_clock_ref_count + 1);
+    });
 }
+#[cfg(any(bt, ieee802154, wifi))]
+fn decrease_phy_clock_ref_count_internal() {
+    critical_section::with(|cs| {
+        let phy_clock_ref_counter = PHY_CLOCK_REF_COUNTERS.borrow(cs);
 
+        let new_phy_clock_ref_count = phy_clock_ref_counter.get().checked_sub(1).expect("PHY clock ref count underflowed. Either you forgot a PhyClockGuard, or used ModemClockController::decrease_phy_clock_ref_count incorrectly.");
```

Comment:
Let use the `unwrap!` macro here. You can still print a message, but it'll be defmt-formatted if that's enabled.